jobs/kpi-forecasting/kpi_forecasting/models/prophet_forecast.py

```diff
@@ -3,65 +3,339 @@
+            "weekly_seasonality": self.weekly_seasonality,
+            "daily_seasonality": self.daily_seasonality,
+            "holidays": self.holidays_raw,
+            "seasonality_mode": self.seasonality_mode,
+            "seasonality_prior_scale": self.seasonality_prior_scale,
+            "holidays_prior_scale": self.holidays_prior_scale,
+            "changepoint_prior_scale": self.changepoint_prior_scale,
+            "mcmc_samples": self.mcmc_samples,
+            "interval_width": self.interval_width,
+            "uncertainty_samples": self.uncertainty_samples,
+            "stan_backend": self.stan_backend,
+            "scaling": self.scaling,
+            "holidays_mode": self.holidays_mode,
+            "cap": self.logistic_growth_cap,
+            "floor": self.logistic_growth_floor,
+            "regressors": self.regressors_raw,
+        }
+
+    @property
+    def column_names_map(self) -> Dict[str, str]:
+        return {"submission_date": "ds", "value": "y"}
+
+    def _add_regressors(self, df: pd.DataFrame, regressors: List[ProphetRegressor]):
+        """
+        Add regressor columns to the dataframe for training or prediction.
+
+        Args:
+            df (pd.DataFrame): The input dataframe.
+            regressors (List[ProphetRegressor]): The list of regressors to add.
+
+        Returns:
+            pd.DataFrame: The dataframe with regressors added.
+        """
+        for regressor in regressors:
+            regressor_time_filter = [True] * len(df)
+            if regressor.start_date:
+                regressor_time_filter &= (
+                    df["ds"] >= pd.to_datetime(regressor.start_date).date()
+                )
+            if regressor.end_date:
+                regressor_time_filter &= (
+                    df["ds"] <= pd.to_datetime(regressor.end_date).date()
+                )
+            # finds rows where date is in regressor date ranges and sets that regressor
+            ## value to 0, else 1
+            df[regressor.name] = (~(regressor_time_filter)).astype(int)
+        return df
+
+    def _set_seed(self) -> None:
+        """Set random seed to ensure that fits and predictions are reproducible."""
+        np.random.seed(42)
+
+    def _build_train_dataframe(self, observed_df) -> pd.DataFrame:
+        """
+        Build the model dataframe for training
+
+        Args:
+            observed_df: dataframe of observed data
+
+        Returns:
+            pd.DataFrame: The dataframe for the model.
+        """
+
+        # define limits for logistic growth
+        observed_df = observed_df.rename(columns=self.column_names_map)
+
+        if self.growth == "logistic":
+            self.logistic_growth_floor = observed_df["y"].min() * 0.5
+            observed_df["floor"] = self.logistic_growth_floor
+            self.logistic_growth_cap = observed_df["y"].max() * 1.5
+            observed_df["cap"] = self.logistic_growth_cap
```

Comment:
Is this used anywhere? These floor/cap scalars seem quite arbitrary, I'd be hesitant to use the scalars as-is.

Reply:
To expand on this a little: logistic growth is convenient for some models because it enables exponential growth at values near the floor and saturating growth at values near the cap. For intermediate regions, the growth rate is roughly linear.

If the distance between the bounds is too large, you lose the dynamic benefits of a logistic curve and most of the growth will be modeled as roughly linear.

Reply:
Yeah it is used in the funnel forecast